File: lib/native/tray.ts
/**
 * Native Tray API
 * TypeScript wrapper for system tray Tauri commands
 */

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import { isTauri } from './utils';
import type { TrayConfig, TrayDisplayMode, TrayState } from '@/types/system/tray';
import { loggers } from '@/lib/logger';

const log = loggers.native;

/**
 * Backend tray config structure (snake_case)
 */
interface BackendTrayConfig {
  display_mode: 'full' | 'compact';
  visible_items: string[];
  item_order: string[];
  show_shortcuts: boolean;
  show_icons: boolean;
  compact_mode_items: string[] | null;
}

/**
 * Backend tray state structure (snake_case)
 */
interface BackendTrayState {
  is_busy: boolean;
  is_recording: boolean;
  status_message: string;
}

/**
 * Convert backend config to frontend format
 */
function toFrontendConfig(config: BackendTrayConfig): TrayConfig {
  return {
    displayMode: config.display_mode,
    visibleItems: config.visible_items,
    itemOrder: config.item_order,
    showShortcuts: config.show_shortcuts,
    showIcons: config.show_icons,
    compactModeItems: config.compact_mode_items ?? undefined,
  };
}

/**
 * Convert frontend config to backend format
 */
function toBackendConfig(config: TrayConfig): BackendTrayConfig {
  return {
    display_mode: config.displayMode,
    visible_items: config.visibleItems,
    item_order: config.itemOrder,
    show_shortcuts: config.showShortcuts,
    show_icons: config.showIcons,
    compact_mode_items: config.compactModeItems ?? null,
  };
}

/**
 * Convert backend state to frontend format
 */
function toFrontendState(state: BackendTrayState): TrayState {
  return {
    isBusy: state.is_busy,
    isRecording: state.is_recording,
    statusMessage: state.status_message,
  };
}

/**
 * Get current tray state
 */
export async function getTrayState(): Promise<TrayState> {
  if (!isTauri()) {
    return {
      isBusy: false,
      isRecording: false,
      statusMessage: '就绪',
    };
  }

  try {
    const state = await invoke<BackendTrayState>('tray_get_state');
    return toFrontendState(state);
  } catch (error) {
    log.error('Failed to get tray state', error as Error);
    throw error;
  }
}

/**
 * Get current tray configuration
 */
export async function getTrayConfig(): Promise<TrayConfig> {
  if (!isTauri()) {
    throw new Error('Tray config is only available in desktop app');
  }

  try {
    const config = await invoke<BackendTrayConfig>('tray_get_config');
    return toFrontendConfig(config);
  } catch (error) {
    log.error('Failed to get tray config', error as Error);
    throw error;
  }
}

/**
 * Set tray configuration
 */
export async function setTrayConfig(config: TrayConfig): Promise<void> {
  if (!isTauri()) {
    throw new Error('Tray config is only available in desktop app');
  }

  try {
    await invoke('tray_set_config', { config: toBackendConfig(config) });
  } catch (error) {
    log.error('Failed to set tray config', error as Error);
    throw error;
  }
}

/**
 * Set tray display mode
 */
export async function setTrayDisplayMode(mode: TrayDisplayMode): Promise<void> {
  if (!isTauri()) {
    throw new Error('Tray is only available in desktop app');
  }

  try {
    await invoke('tray_set_display_mode', { mode });
  } catch (error) {
    log.error('Failed to set tray display mode', error as Error);
    throw error;
  }
}

/**
 * Toggle tray display mode between full and compact
 */
export async function toggleTrayDisplayMode(): Promise<TrayDisplayMode> {
  if (!isTauri()) {
    throw new Error('Tray is only available in desktop app');
  }

  try {
    return await invoke<TrayDisplayMode>('tray_toggle_display_mode');
  } catch (error) {
    log.error('Failed to toggle tray display mode', error as Error);
    throw error;
  }
}

/**
 * Set item visibility in tray menu
 */
export async function setTrayItemVisibility(
  itemId: string,
  visible: boolean
): Promise<void> {
  if (!isTauri()) {
    throw new Error('Tray is only available in desktop app');
  }

  try {
    await invoke('tray_set_item_visibility', { itemId, visible });
  } catch (error) {
    log.error('Failed to set tray item visibility', error as Error);
    throw error;
  }
}

/**
 * Set compact mode items
 */
export async function setTrayCompactItems(items: string[]): Promise<void> {
  if (!isTauri()) {
    throw new Error('Tray is only available in desktop app');
  }

  try {
    await invoke('tray_set_compact_items', { items });
  } catch (error) {
    log.error('Failed to set compact items', error as Error);
    throw error;
  }
}

/**
 * Update tray tooltip message
 */
export async function updateTrayTooltip(message: string): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('tray_update_tooltip', { message });
  } catch (error) {
    log.error('Failed to update tray tooltip', error as Error);
    throw error;
  }
}

/**
 * Set tray busy state
 */
export async function setTrayBusy(
  busy: boolean,
  message?: string
): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('tray_set_busy', { busy, message: message ?? null });
  } catch (error) {
    log.error('Failed to set tray busy state', error as Error);
    throw error;
  }
}

/**
 * Refresh tray menu
 */
export async function refreshTrayMenu(): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('tray_refresh_menu');
  } catch (error) {
    log.error('Failed to refresh tray menu', error as Error);
    throw error;
  }
}

/**
 * Get default compact mode items
 */
export async function getDefaultCompactItems(): Promise<string[]> {
  if (!isTauri()) {
    return [
      'toggle-chat-widget',
      'screenshot-region',
      'clipboard-history',
      'open-settings',
      'quit',
    ];
  }

  try {
    return await invoke<string[]>('tray_get_default_compact_items');
  } catch (error) {
    log.error('Failed to get default compact items', error as Error);
    throw error;
  }
}

/**
 * Get all available menu item IDs
 */
export async function getAllTrayItemIds(): Promise<string[]> {
  if (!isTauri()) {
    return [];
  }

  try {
    return await invoke<string[]>('tray_get_all_item_ids');
  } catch (error) {
    log.error('Failed to get all tray item IDs', error as Error);
    throw error;
  }
}

/**
 * Listen for tray config changes
 */
export async function onTrayConfigChanged(
  callback: () => void
): Promise<UnlistenFn> {
  if (!isTauri()) {
    return () => {};
  }

  return listen('tray-config-changed', callback);
}

/**
 * Listen for tray state changes
 */
export async function onTrayStateChanged(
  callback: (state: TrayState) => void
): Promise<UnlistenFn> {
  if (!isTauri()) {
    return () => {};
  }

  return listen<BackendTrayState>('tray-state-changed', (event) => {
    callback(toFrontendState(event.payload));
  });
}

File: src-tauri/src/commands/system/tray.rs
//! Tray Commands
//!
//! Commands for system tray configuration and control.

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, State};

use crate::tray::{refresh_tray_menu, set_tray_busy, update_tray_tooltip, TRAY_STATE};

/// Tray display mode
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TrayDisplayMode {
    Full,
    Compact,
}

/// Tray configuration from frontend
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrayConfig {
    pub display_mode: TrayDisplayMode,
    pub visible_items: Vec<String>,
    pub item_order: Vec<String>,
    pub show_shortcuts: bool,
    pub show_icons: bool,
    pub compact_mode_items: Option<Vec<String>>,
}

/// Tray state response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrayStateResponse {
    pub is_busy: bool,
    pub is_recording: bool,
    pub status_message: String,
}

/// Managed tray configuration state
pub struct TrayConfigState {
    pub config: parking_lot::RwLock<TrayConfig>,
}

impl Default for TrayConfigState {
    fn default() -> Self {
        Self {
            config: parking_lot::RwLock::new(TrayConfig {
                display_mode: TrayDisplayMode::Full,
                visible_items: vec![],
                item_order: vec![],
                show_shortcuts: true,
                show_icons: true,
                compact_mode_items: None,
            }),
        }
    }
}

/// Get current tray state
#[tauri::command]
pub fn tray_get_state() -> TrayStateResponse {
    let state = TRAY_STATE.read();
    TrayStateResponse {
        is_busy: state.is_busy,
        is_recording: state.is_recording,
        status_message: state.status_message.clone(),
    }
}

/// Get current tray configuration
#[tauri::command]
pub fn tray_get_config(state: State<'_, TrayConfigState>) -> TrayConfig {
    state.config.read().clone()
}

/// Set tray configuration
#[tauri::command]
pub fn tray_set_config(
    app: AppHandle,
    state: State<'_, TrayConfigState>,
    config: TrayConfig,
) -> Result<(), String> {
    // Update stored config
    {
        let mut stored = state.config.write();
        *stored = config;
    }

    // Refresh the tray menu
    refresh_tray_menu(&app);

    // Emit event to notify frontend
    let _ = app.emit("tray-config-changed", ());

    log::info!("Tray configuration updated");
    Ok(())
}

/// Set tray display mode
#[tauri::command]
pub fn tray_set_display_mode(
    app: AppHandle,
    state: State<'_, TrayConfigState>,
    mode: TrayDisplayMode,
) -> Result<(), String> {
    {
        let mut config = state.config.write();
        config.display_mode = mode;
    }

    // Refresh the tray menu
    refresh_tray_menu(&app);

    log::info!("Tray display mode set to {:?}", mode);
    Ok(())
}

/// Toggle tray display mode between full and compact
#[tauri::command]
pub fn tray_toggle_display_mode(
    app: AppHandle,
    state: State<'_, TrayConfigState>,
) -> Result<TrayDisplayMode, String> {
    let new_mode = {
        let mut config = state.config.write();
        config.display_mode = match config.display_mode {
            TrayDisplayMode::Full => TrayDisplayMode::Compact,
            TrayDisplayMode::Compact => TrayDisplayMode::Full,
        };
        config.display_mode
    };

    // Refresh the tray menu
    refresh_tray_menu(&app);

    log::info!("Tray display mode toggled to {:?}", new_mode);
    Ok(new_mode)
}

/// Set item visibility
#[tauri::command]
pub fn tray_set_item_visibility(
    app: AppHandle,
    state: State<'_, TrayConfigState>,
    item_id: String,
    visible: bool,
) -> Result<(), String> {
    {
        let mut config = state.config.write();
        if visible {
            if !config.visible_items.contains(&item_id) {
                config.visible_items.push(item_id.clone());
            }
        } else {
            config.visible_items.retain(|id| id != &item_id);
        }
    }

    // Refresh the tray menu
    refresh_tray_menu(&app);

    log::debug!("Tray item '{}' visibility set to {}", item_id, visible);
    Ok(())
}

/// Set compact mode items
#[tauri::command]
pub fn tray_set_compact_items(
    app: AppHandle,
    state: State<'_, TrayConfigState>,
    items: Vec<String>,
) -> Result<(), String> {
    {
        let mut config = state.config.write();
        config.compact_mode_items = Some(items);
    }

    // Refresh the tray menu if in compact mode
    let is_compact = {
        let config = state.config.read();
        config.display_mode == TrayDisplayMode::Compact
    };

    if is_compact {
        refresh_tray_menu(&app);
    }

    log::info!("Compact mode items updated");
    Ok(())
}

/// Update tray tooltip
#[tauri::command]
pub fn tray_update_tooltip(app: AppHandle, message: String) -> Result<(), String> {
    update_tray_tooltip(&app, &message);
    Ok(())
}

/// Set tray busy state
#[tauri::command]
pub fn tray_set_busy(app: AppHandle, busy: bool, message: Option<String>) -> Result<(), String> {
    set_tray_busy(&app, busy, message.as_deref());
    Ok(())
}

/// Refresh tray menu
#[tauri::command]
pub fn tray_refresh_menu(app: AppHandle) -> Result<(), String> {
    refresh_tray_menu(&app);
    Ok(())
}

/// Get default compact mode items
#[tauri::command]
pub fn tray_get_default_compact_items() -> Vec<String> {
    vec![
        "toggle-chat-widget".to_string(),
        "screenshot-region".to_string(),
        "clipboard-history".to_string(),
        "open-settings".to_string(),
        "quit".to_string(),
    ]
}

/// Get all available menu item IDs
#[tauri::command]
pub fn tray_get_all_item_ids() -> Vec<String> {
    vec![
        // Window controls
        "show-window".to_string(),
        "hide-window".to_string(),
        "toggle-chat-widget".to_string(),
        "bubble-show".to_string(),
        "bubble-hide".to_string(),
        "bubble-toggle-minimize".to_string(),
        // Screenshot
        "screenshot-menu".to_string(),
        "screenshot-fullscreen".to_string(),
        "screenshot-region".to_string(),
        "screenshot-window".to_string(),
        "screenshot-ocr".to_string(),
        // Recording
        "recording-menu".to_string(),
        "recording-start".to_string(),
        "recording-stop".to_string(),
        "recording-pause".to_string(),
        // Selection
        "selection-menu".to_string(),
        "selection-enabled".to_string(),
        "selection-trigger".to_string(),
        "selection-hide-toolbar".to_string(),
        "selection-restart".to_string(),
        // Clipboard
        "clipboard-menu".to_string(),
        "clipboard-history".to_string(),
        "clipboard-clear".to_string(),
        // Settings
        "settings-menu".to_string(),
        "autostart-enabled".to_string(),
        "open-settings".to_string(),
        "open-logs".to_string(),
        // Help
        "check-update".to_string(),
        "open-help".to_string(),
        "about".to_string(),
        // Exit
        "quit".to_string(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tray_display_mode_serialize() {
        assert_eq!(
            serde_json::to_string(&TrayDisplayMode::Full).unwrap(),
            "\"full\""
        );
        assert_eq!(
            serde_json::to_string(&TrayDisplayMode::Compact).unwrap(),
            "\"compact\""
        );
    }

    #[test]
    fn test_tray_config_state_default() {
        let state = TrayConfigState::default();
        let config = state.config.read();
        assert_eq!(config.display_mode, TrayDisplayMode::Full);
        assert!(config.show_shortcuts);
        assert!(config.show_icons);
    }

    #[test]
    fn test_get_default_compact_items() {
        let items = tray_get_default_compact_items();
        assert_eq!(items.len(), 5);
        assert!(items.contains(&"toggle-chat-widget".to_string()));
        assert!(items.contains(&"quit".to_string()));
    }

    #[test]
    fn test_get_all_item_ids() {
        let items = tray_get_all_item_ids();
        assert!(items.len() > 20);
        assert!(items.contains(&"show-window".to_string()));
        assert!(items.contains(&"quit".to_string()));
    }
}
